# ğŸ“ ATC202: Multi parameters should be separated onto individual lines

## ğŸ“‚ Category

Style

## âš ï¸ Severity

Warning

## ğŸ“– Description

Ensures consistent parameter formatting across method declarations. Parameters should be on separate lines for methods with multiple parameters or when the total line length exceeds 80 characters.

## ğŸ“‹ Rules

1. **Multiple parameters (2 or more)**: All parameters must be on separate lines
2. **Single parameter**: Must be on a separate line if the total line length exceeds 80 characters
3. **Applies to**: Method declarations, constructor declarations, local functions, and delegate declarations
4. **Does NOT apply to**: Method invocations, lambda expressions, ternary operators, LINQ queries, or indexers

## ğŸ’¡ Motivation

Consistent parameter formatting improves code readability and makes it easier to:
- Review parameter lists in pull requests
- Understand method signatures at a glance
- Maintain consistent code style across the codebase
- Avoid horizontal scrolling in code editors

## âŒ Bad Code

### Multiple parameters on one line

```csharp
public void ProcessOrder(int orderId, string customerName, decimal amount)
{
    // Implementation
}
```

### Single parameter on long line (exceeds 80 characters)

```csharp
public void MyVeryLongMethodNameThatExceedsTheLineLimit(string parameterName)
{
    // Implementation
}
```

### Constructor with multiple parameters on one line

```csharp
public class Customer
{
    public Customer(string name, string email, string phone)
    {
        // Implementation
    }
}
```

## âœ… Good Code

### Multiple parameters on separate lines

```csharp
public void ProcessOrder(
    int orderId,
    string customerName,
    decimal amount)
{
    // Implementation
}
```

### Single parameter on separate line when line is long

```csharp
public void MyVeryLongMethodNameThatExceedsTheLineLimit(
    string parameterName)
{
    // Implementation
}
```

### Constructor with parameters on separate lines

```csharp
public class Customer
{
    public Customer(
        string name,
        string email,
        string phone)
    {
        // Implementation
    }
}
```

### Single parameter on same line (line is short)

```csharp
public void Process(int id)
{
    // Implementation
}
```

## âš™ï¸ Configuration

The maximum line length threshold is set to 80 characters by default and is currently not configurable.

## ğŸ¤– Generated Code

This rule automatically skips analysis of generated code. Generated code is identified by:

1. **GeneratedCode Attribute**: Classes or types marked with `[GeneratedCode]` attribute from `System.CodeDom.Compiler`
2. **Auto-generated Headers**: Files containing "auto-generated" in header comments (case-insensitive)

Example of skipped code:

```csharp
using System.CodeDom.Compiler;

[GeneratedCode("MyCodeGenerator", "1.0")]
public class GeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

```csharp
//------------------------------------------------------------------------------
// This code was auto-generated by ApiGenerator 2.0.
//
// Changes to this file may cause incorrect behavior and will be lost if
// the code is regenerated.
//------------------------------------------------------------------------------

public class AutoGeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

## ğŸ”§ Code Fix

A code fix is available that automatically formats parameters onto separate lines with proper indentation.

## ğŸ”— Related Rules

- [ATC201: Single parameter should be kept inline when declaration is short](ATC201.md) - Complementary rule that enforces inline formatting for single-parameter methods when the line is short
