# üìè ATC202: Multi parameters should be separated on individual lines

## üìÇ Category

Style

## ‚ö†Ô∏è Severity

Warning

## üìñ Description

Ensures consistent parameter formatting for methods with multiple parameters. All parameters must be placed on separate lines to improve readability and make code reviews easier.

## üìã Rules

1. **Multiple parameters (2 or more)**: All parameters must be on separate lines
2. **Applies to**: Method declarations, constructor declarations, local functions, and delegate declarations with 2 or more parameters
3. **Does NOT apply to**:
   - Single-parameter methods (handled by ATC201)
   - Method invocations
   - Lambda expressions
   - Ternary operators
   - LINQ queries
   - Indexers

## üí° Motivation

Consistent parameter formatting improves code readability and makes it easier to:

- Review parameter lists in pull requests
- Understand method signatures at a glance
- Maintain consistent code style across the codebase
- Avoid horizontal scrolling in code editors

## üìù Examples

### Multiple parameters on same line

```csharp
// non-compliant - all parameters on same line
public void ProcessOrder(int orderId, string customerName, decimal amount)
{
    // Implementation
}

// compliant - each parameter on separate line
public void ProcessOrder(
    int orderId,
    string customerName,
    decimal amount)
{
    // Implementation
}
```

### Constructor with multiple parameters

```csharp
// non-compliant
public class Customer
{
    public Customer(string name, string email, string phone)
    {
        // Implementation
    }
}

// compliant
public class Customer
{
    public Customer(
        string name,
        string email,
        string phone)
    {
        // Implementation
    }
}
```

### Method with two parameters

```csharp
// non-compliant
public void Update(int id, string value)
{
    // Implementation
}

// compliant
public void Update(
    int id,
    string value)
{
    // Implementation
}
```

### Delegate with multiple parameters

```csharp
// non-compliant
public delegate void ProcessHandler(object sender, EventArgs e, string message);

// compliant
public delegate void ProcessHandler(
    object sender,
    EventArgs e,
    string message);
```

## ‚öôÔ∏è Configuration

Configure the minimum parameter count threshold in `.editorconfig`:

```ini
[*.cs]
# Minimum number of parameters before requiring separation (default: 2, range: 2-10)
dotnet_diagnostic.ATC202.min_parameter_count = 3
```

| Option | Default | Range | Description |
|--------|---------|-------|-------------|
| `min_parameter_count` | 2 | 2-10 | Minimum number of parameters before the rule requires separation |

**Example configurations:**

```ini
# Default - 2+ parameters must be on separate lines
dotnet_diagnostic.ATC202.min_parameter_count = 2

# Lenient - only 3+ parameters must be on separate lines
dotnet_diagnostic.ATC202.min_parameter_count = 3

# Strict - only 4+ parameters must be on separate lines
dotnet_diagnostic.ATC202.min_parameter_count = 4
```

## ü§ñ Generated Code

This rule automatically skips analysis of generated code. Generated code is identified by:

1. **GeneratedCode Attribute**: Classes or types marked with `[GeneratedCode]` attribute from `System.CodeDom.Compiler`
2. **Auto-generated Headers**: Files containing "auto-generated" in header comments (case-insensitive)

Example of skipped code:

```csharp
using System.CodeDom.Compiler;

[GeneratedCode("MyCodeGenerator", "1.0")]
public class GeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

```csharp
//------------------------------------------------------------------------------
// This code was auto-generated by ApiGenerator 2.0.
//
// Changes to this file may cause incorrect behavior and will be lost if
// the code is regenerated.
//------------------------------------------------------------------------------

public class AutoGeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

## üîß Code Fix

A code fix is available that automatically formats parameters on separate lines with proper indentation.

**How to apply:**

1. Position cursor on the method/constructor with multiple parameters on one line
2. Click the lightbulb (üí°) or press `Ctrl+.` (Windows/Linux) or `Cmd+.` (Mac)
3. Select: **"Separate multi parameters on individual lines"**

The code fix will:

- Place each parameter on its own line
- Apply consistent indentation (4 spaces per level)
- Preserve parameter attributes, modifiers, and default values
- Maintain proper spacing and line breaks

**Example:**

```csharp
// Before code fix
public void ProcessOrder(int orderId, string customerName, decimal amount)

// After code fix (automatic)
public void ProcessOrder(
    int orderId,
    string customerName,
    decimal amount)
```

**Supported declarations:**

- Methods
- Constructors (regular, not primary constructors)
- Local functions
- Delegates

**Note:** Primary constructors (C# 12+) for classes and records are not currently supported by this rule.

## üîÄ Overlap with StyleCop Rules

This rule has functional overlap with StyleCop Analyzers rules:

| StyleCop Rule | Description | Comparison |
|---------------|-------------|------------|
| [SA1116](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1116.md) | Split parameters must start on line after declaration | Both have code fixes |
| [SA1117](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1117.md) | Parameters must be on same line or separate lines | Both have code fixes |

**Key advantage of ATC202:** This rule provides **configurable minimum parameter count** via `.editorconfig`, allowing teams to adjust when the rule applies (default 2 parameters, range 2-10).

### Configuration recommendation

If you enable ATC202, consider disabling the overlapping StyleCop rules to avoid duplicate diagnostics:

```editorconfig
[*.cs]
# Disable StyleCop rules that overlap with ATC201/ATC202
dotnet_diagnostic.SA1116.severity = none
dotnet_diagnostic.SA1117.severity = none

# Enable ATC202 with custom minimum parameter count
dotnet_diagnostic.ATC202.severity = warning
dotnet_diagnostic.ATC202.min_parameter_count = 2
```

## üîó Related Rules

- [ATC201: Single parameter should be kept inline when declaration is short](ATC201.md) - Parameter formatting for single parameters
- [ATC203: Method chains with 2 or more calls should be placed on separate lines](ATC203.md) - Method chain formatting
- [ATC210: Use expression body syntax when appropriate](ATC210.md) - Expression body formatting
- [ATC230: Require exactly one blank line between code blocks](ATC230.md) - Blank line formatting
