# ğŸ“ ATC202: Multi parameters should be separated onto individual lines

## ğŸ“‚ Category

Style

## âš ï¸ Severity

Warning

## ğŸ“– Description

Ensures consistent parameter formatting for methods with multiple parameters. All parameters must be placed on separate lines to improve readability and make code reviews easier.

## ğŸ“‹ Rules

1. **Multiple parameters (2 or more)**: All parameters must be on separate lines
2. **Applies to**: Method declarations, constructor declarations, local functions, and delegate declarations with 2 or more parameters
3. **Does NOT apply to**:
   - Single-parameter methods (handled by ATC201)
   - Method invocations
   - Lambda expressions
   - Ternary operators
   - LINQ queries
   - Indexers

## ğŸ’¡ Motivation

Consistent parameter formatting improves code readability and makes it easier to:
- Review parameter lists in pull requests
- Understand method signatures at a glance
- Maintain consistent code style across the codebase
- Avoid horizontal scrolling in code editors

## ğŸ“ Examples

### Multiple parameters on same line

```csharp
// non-compliant - all parameters on same line
public void ProcessOrder(int orderId, string customerName, decimal amount)
{
    // Implementation
}

// compliant - each parameter on separate line
public void ProcessOrder(
    int orderId,
    string customerName,
    decimal amount)
{
    // Implementation
}
```

### Constructor with multiple parameters

```csharp
// non-compliant
public class Customer
{
    public Customer(string name, string email, string phone)
    {
        // Implementation
    }
}

// compliant
public class Customer
{
    public Customer(
        string name,
        string email,
        string phone)
    {
        // Implementation
    }
}
```

### Method with two parameters

```csharp
// non-compliant
public void Update(int id, string value)
{
    // Implementation
}

// compliant
public void Update(
    int id,
    string value)
{
    // Implementation
}
```

### Delegate with multiple parameters

```csharp
// non-compliant
public delegate void ProcessHandler(object sender, EventArgs e, string message);

// compliant
public delegate void ProcessHandler(
    object sender,
    EventArgs e,
    string message);
```

## âš™ï¸ Configuration

This rule has no configurable options. It always requires multiple parameters (2 or more) to be on separate lines.

## ğŸ¤– Generated Code

This rule automatically skips analysis of generated code. Generated code is identified by:

1. **GeneratedCode Attribute**: Classes or types marked with `[GeneratedCode]` attribute from `System.CodeDom.Compiler`
2. **Auto-generated Headers**: Files containing "auto-generated" in header comments (case-insensitive)

Example of skipped code:

```csharp
using System.CodeDom.Compiler;

[GeneratedCode("MyCodeGenerator", "1.0")]
public class GeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

```csharp
//------------------------------------------------------------------------------
// This code was auto-generated by ApiGenerator 2.0.
//
// Changes to this file may cause incorrect behavior and will be lost if
// the code is regenerated.
//------------------------------------------------------------------------------

public class AutoGeneratedClass
{
    // This would normally violate ATC202, but is ignored
    public void Method(int x, string y, bool z)
    {
    }
}
```

## ğŸ”§ Code Fix

A code fix is available that automatically formats parameters onto separate lines with proper indentation.

## ğŸ”— Related Rules

- [ATC201: Single parameter should be kept inline when declaration is short](ATC201.md) - Complementary rule that enforces inline formatting for single-parameter methods when the line is short
