# ğŸ“ ATC201: Single parameter should be kept inline when declaration is short

## ğŸ“‚ Category

Style

## âš ï¸ Severity

Warning

## ğŸ“– Description

Enforces that single-parameter methods keep the parameter on the same line as the method declaration when the total line length would be under 80 characters. This rule ensures consistent and readable code by avoiding unnecessary line breaks for short method signatures.

## ğŸ“‹ Rules

1. **Single parameter methods**: If the entire declaration (including parameter) would fit within 80 characters, the parameter must remain on the same line
2. **Long declarations**: If the declaration would exceed 80 characters, breaking the parameter to a new line is allowed (and required by this rule's complement)
3. **Applies to**: Method declarations, constructor declarations, local functions, and delegate declarations
4. **Does NOT apply to**: Methods with multiple parameters (handled by ATC202), method invocations, lambda expressions

## ğŸ’¡ Motivation

Keeping short single-parameter declarations on one line:
- Improves code readability by reducing unnecessary vertical space
- Makes method signatures more scannable
- Maintains consistency with common C# conventions
- Reduces visual clutter in codebases

This rule works in tandem with ATC202 to provide comprehensive parameter formatting guidance.

## âŒ Bad Code

### Single short parameter unnecessarily broken to new line

```csharp
public void Process(
    int id)
{
    // Implementation
}
```

### Short constructor with parameter on new line

```csharp
public class Customer
{
    public Customer(
        string name)
    {
        // Implementation
    }
}
```

### Short method with nullable parameter on new line

```csharp
public void Handle(
    string? message)
{
    // Implementation
}
```

## âœ… Good Code

### Single parameter on same line (short declaration)

```csharp
public void Process(int id)
{
    // Implementation
}
```

### Constructor with parameter inline

```csharp
public class Customer
{
    public Customer(string name)
    {
        // Implementation
    }
}
```

### Method with nullable parameter inline

```csharp
public void Handle(string? message)
{
    // Implementation
}
```

### Long declaration with parameter on new line (allowed)

```csharp
// This is OK because the line would exceed 80 characters
public void MyVeryLongMethodNameThatExceedsTheLineLimitWithParameter(
    int parameter)
{
    // Implementation
}
```

## âš™ï¸ Configuration

The maximum line length threshold is set to 80 characters by default and is currently not configurable.

## ğŸ¤– Generated Code

This rule automatically skips analysis of generated code. Generated code is identified by:

1. **GeneratedCode Attribute**: Classes or types marked with `[GeneratedCode]` attribute from `System.CodeDom.Compiler`
2. **Auto-generated Headers**: Files containing "auto-generated" in header comments (case-insensitive)

Example of skipped code:

```csharp
using System.CodeDom.Compiler;

[GeneratedCode("MyCodeGenerator", "1.0")]
public class GeneratedClass
{
    // This would normally violate ATC201, but is ignored
    public void Method(
        int parameter)
    {
    }
}
```

```csharp
//------------------------------------------------------------------------------
// This code was auto-generated by ApiGenerator 2.0.
//
// Changes to this file may cause incorrect behavior and will be lost if
// the code is regenerated.
//------------------------------------------------------------------------------

public class AutoGeneratedClass
{
    // This would normally violate ATC201, but is ignored
    public void Method(
        int parameter)
    {
    }
}
```

## ğŸ”§ Code Fix

Currently, no automatic code fix is available for this rule. Manual correction is required.

## ğŸ”— Related Rules

- [ATC202: Multi parameters should be separated onto individual lines](ATC202.md) - Complementary rule that enforces parameter separation for multiple parameters or long lines
